README.rst
====================================================
 galaxy-upload - Galaxy Command-Line Upload Utility
====================================================

A utility for uploading files to a Galaxy server from the command line. galaxy-upload supports `Galaxy`_ servers 22.01+,
which include support for resumable uploads with the `tus`_ protocol.

Installation
============

Using pip
---------

Python 3.7 or later is required.

To install::

    $ pip install galaxy-upload

This installs two commands: ``galaxy-upload``, used to upload file(s) to a Galaxy server, and ``galaxy-history-search``,
a helper utility for finding Galaxy histories to pass to the ``galaxy-upload`` command.

To make your life easier, you are encourged to install into a Python virtual environment. The easiest way to do this is
with Python's built-in `venv`_ module::

    $ python3 -m venv ~/galaxy-upload
    $ . ~/galaxy-upload/bin/activate
    $ pip install galaxy-upload

Using Conda
-----------

Alternatively, galaxy-upload can be installed using the `Conda`_ package manager. The `galaxy-upload Conda package`_ can
be found on the `bioconda`_ channel and installed like so::

    $ conda create -n galaxy-upload -c conda-forge -c bioconda galaxy-upload
    $ conda activate galaxy-upload

Using Containers
----------------

It is also possible to run galaxy-upload in either a `Docker`_ or `Singularity`_/`Apptainer`_ container. The
`galaxy-upload BioContainer`_ is automatically built and maintained by the `BioContainers`_ project.

To use the Docker container::

    $ docker run -it --rm -v "$(pwd):$(pwd)" -w "$(pwd)" -u "$(id -u):$(id -g)" \
        quay.io/biocontainers/galaxy-upload:1.0.0--pyhdfd78af_0 /bin/bash

Or as a single command without entering an interactive shell::

    $ docker run --rm -v "$(pwd):$(pwd)" -w "$(pwd)" -u "$(id -u):$(id -g)" \
        quay.io/biocontainers/galaxy-upload:1.0.0--pyhdfd78af_0 galaxy-upload

Adjust the values of ``-v`` and ``-w`` according to where the data you want to upload are located. In the example above,
it is assumed they are in the current working directory.

To use the Singularity container::

    $ singularity run https://depot.galaxyproject.org/singularity/galaxy-upload:1.0.0--pyhdfd78af_0

Or as a single command without entering an interactive shell::

    $ singularity run https://depot.galaxyproject.org/singularity/galaxy-upload:1.0.0--pyhdfd78af_0 galaxy-upload

Additional (newer) versions of the container may be available, BioContainers does not use the ``latest`` tag, but you
can find all tags (which are valid for the Singularity images hosted on depot.galaxyproject.org as well as the Docker
images) at the `galaxy-upload quay.io page`_

Usage
=====

Upload a pair of fastq.gz files::

    $ galaxy-upload --url https://usegalaxy.org \
        --api-key 70ffeec0ffeea11e1eaccede1337loaf --history-name 'Run 2' \
        RUN2_F_001.fastq.gz RUN2_R_001.fastq.gz
    RUN2_F_001.fastq.gz ━━━━━━━━━━━━━━━━━━━━━━━━━━ 100/100 mB ? eta 0:00:00
    RUN2_R_001.fastq.gz ━━━━━━━━━━━━━━━━━━━━━━━━━━ 100/100 mB ? eta 0:00:00

Required arguments are the Galaxy server URL and API key, and a history ID or name. Your API key can be found in the
Galaxy UI after logging in, by navigating to **User** ⮕ **Preferences** ⮕ **Manage API Key**.

You can set the URL and API key options as environment variables to avoid retyping and prevent the key from being
visible to other users in ``ps(1)`` output::

    $ export GALAXY_URL=https://usegalaxy.org 
    $ export GALAXY_API_KEY=70ffeec0ffeea11e1eaccede1337loaf
    $ galaxy-upload --history-name stuff reads.bam

When selecting a history by name, regular expression matching is used. If the name matches multiple histories,
``galaxy-upload`` will exit will output details about the matched histories and then exit with an error. You can then
select the correct history ID using the ``--history-id`` option::

    $ galaxy-upload --history-name stuff reads.bam
                              Active Histories
    ┏━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━┓
    ┃ ID               ┃ Name        ┃ Last Modified            ┃
    ┡━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━┩
    │ 70ffeec0ffeead07 │ Newer stuff │ Mon Jul 11 15:54:05 2022 │
    │ a11e1eaccedeble8 │ Older stuff │ Wed May 25 18:03:46 2022 │
    └──────────────────┴─────────────┴──────────────────────────┘
    ERROR: Multiple histories matching stuff found! Use --history-id to select one.
    $ galaxy-upload --history-id 70ffeec0ffeead07 reads.bam
    reads.bam ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.6/3.6 gB ? eta 0:00:00

If you want to find the correct history without attempting an upload, use the ``galaxy-history-search`` command. The
``--ignore-case`` option can be used to perform a case-insensitive search::

    $ galaxy-history-search --ignore-case trinity
                               Active Histories
    ┏━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━┓
    ┃ ID               ┃ Name                  ┃ Last Modified            ┃
    ┡━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━┩
    │ 084649feb42d4295 │ : test trinity inputs │ Wed Dec  9 10:02:35 2020 │
    │ f697f94ca47080cf │ automate_trinity      │ Mon Dec 21 17:40:24 2015 │
    │ c79278c7a37e619e │ TrinityRun            │ Fri Mar 10 14:21:56 2017 │
    │ ee31286b26ff3352 │ trinity               │ Wed Sep 30 09:04:03 2020 │
    └──────────────────┴───────────────────────┴──────────────────────────┘

Regular expressions are supported, for example, to find only the histories with names *ending* with ``trinity``::

    $ galaxy-history-search --ignore-case 'trinity$'
                             Active Histories
    ┏━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━┓
    ┃ ID               ┃ Name             ┃ Last Modified            ┃
    ┡━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━┩
    │ f697f94ca47080cf │ automate_trinity │ Mon Dec 21 17:40:24 2015 │
    │ ee31286b26ff3352 │ trinity          │ Wed Sep 30 09:04:03 2020 │
    └──────────────────┴──────────────────┴──────────────────────────┘

Multiple options mirror those of the Galaxy UI's upload dialog, including ``--file-type``, ``--dbkey``,
and ``--space-to-tab``. The ``--file-name`` option can be used when uploading single files to control the file name in
the history (by default it will be the same as the name on the local filesystem).

To support resuming interrupted uploads, use the ``--storage`` option to point to a state file (it will be created if it
does not exist)::

    $ galaxy-upload --file-type bam --file-name Reads --storage /data/upload.txt /data/reads.bam

If the upload is interrupted, simply repeat the same command to resume uploading from the point of interruption.

Note that if you are trying to re-upload (not resume) a file that you have already uploaded once before, you will need
to remove it from the storage file or use a different storage file.

.. _Galaxy: http://galaxyproject.org/
.. _tus: https://tus.io/
.. _venv: https://docs.python.org/3/library/venv.html
.. _Conda: https://docs.conda.io/
.. _galaxy-upload Conda package: https://anaconda.org/bioconda/galaxy-upload
.. _bioconda: https://bioconda.github.io/
.. _Docker: https://www.docker.com/
.. _Singularity: https://sylabs.io/docs/
.. _Apptainer: https://apptainer.org/
.. _galaxy-upload BioContainer: https://biocontainers.pro/tools/galaxy-upload
.. _BioContainers: https://biocontainers.pro/
.. _galaxy-upload quay.io page: https://quay.io/repository/biocontainers/galaxy-upload


HISTORY.rst
=========
 History
=========

1.0.1
=====

- Document installation/usage with conda and containers by @natefoo in https://github.com/galaxyproject/galaxy-upload/pull/7
- Use click to require file arg to exist by @natefoo in https://github.com/galaxyproject/galaxy-upload/pull/9
- Pin tuspy to !=1.0.1 due to issue with upload_chunk(). by @natefoo in https://github.com/galaxyproject/galaxy-upload/pull/10

1.0.0
=====

- Initial release

0.0.1
=====

- Initial prerelease for comment and testing


gxupload/main.py
#!/usr/bin/env python3
"""Upload files to a Galaxy instance."""
import os
import sys
import traceback

import click
import rich.console
import rich.progress
from click_option_group import optgroup, RequiredMutuallyExclusiveOptionGroup
from tusclient.fingerprint import fingerprint

from bioblend import ConnectionError
from bioblend.galaxy import GalaxyInstance

from .history import get_histories, make_table


def make_bar(file_name, total=None):
    columns = (
        rich.progress.TextColumn("[progress.description]{task.description}"),
        rich.progress.BarColumn(),
        rich.progress.DownloadColumn(),
        rich.progress.TransferSpeedColumn(),
        rich.progress.TextColumn("eta"),
        rich.progress.TimeRemainingColumn(),
    )

    bar = rich.progress.Progress(*columns)
    task_id = bar.add_task(file_name, total=total)
    return (bar, task_id)


def find_history(gi, history_name, ignore_case):
    histories = get_histories(gi, ignore_case=ignore_case, name=history_name)
    if not histories:
        message = f"ERROR: No histories matching {history_name} found!"
        click.echo(click.style(message, bold=True, fg="red"), err=True)
        sys.exit(1)
    elif len(histories) > 1:
        table = make_table(quiet=False, histories=histories)
        console = rich.console.Console()
        console.print(table)
        message = f"ERROR: Multiple histories matching {history_name} found! Use --history-id to select one."
        click.echo(click.style(message, bold=True, fg="red"), err=True)
        sys.exit(1)
    return histories[0]


def upload_file(
    url,
    path,
    api_key,
    history_id=None,
    history_name=None,
    ignore_case=False,
    file_type="auto",
    dbkey="?",
    space_to_tab=False,
    auto_decompress=False,
    file_name=None,
    storage=None,
    silent=False,
    debug=False,
):
    file_name = file_name or os.path.basename(path)

    gi = GalaxyInstance(url, api_key)

    upload_kwargs = {
        "file_type": file_type,
        "dbkey": dbkey,
        "file_name": file_name,
        "space_to_tab": space_to_tab,
        "auto_decompress": auto_decompress,
    }

    if not history_id:
        history = find_history(gi, history_name, ignore_case)
        history_id = history["id"]
        if debug:
            click.echo(f"History name search '{history_name}' resolved to history {history_id}: {history['name']}")

    try:
        if silent:
            gi.tools.upload_file_tus(
                path,
                history_id,
                storage=storage,
                **upload_kwargs,
            )
            return

        uploader = gi.get_tus_uploader(
            path,
            storage=storage,
        )
        # TODO: this is uploader.get_file_size() in tusclient 1.0.0
        bar, task_id = make_bar(file_name, total=uploader.file_size)

        with bar:
            last_offset = 0
            while uploader.offset < uploader.file_size:
                uploader.upload_chunk()
                bar.update(task_id, advance=(uploader.offset - last_offset))
                last_offset = uploader.offset

        gi.tools.post_to_fetch(path, history_id, uploader.session_id, **upload_kwargs)

    except ConnectionError as exc:
        if exc.status_code == 404 and storage:
            fingerprinter = fingerprint.Fingerprint()
            with open(path, "rb") as fh:
                fp_hash = fingerprinter.get_fingerprint(fh)
            message = (
                f"Unable to resume, previous upload may have been removed from server (hint: remove {fp_hash}"
                f" from {storage} or change storage to reupload from the start: {exc}"
            )
        else:
            message = str(exc)
        message = f"ERROR: {message}"
        if debug:
            click.echo(traceback.format_exc(), nl=False)
        click.echo(click.style(message, bold=True, fg="red"), err=True)


@click.command()
@optgroup.group("History Selection Options", cls=RequiredMutuallyExclusiveOptionGroup)
@optgroup.option("--history-id", type=str, help="Target history ID")
@optgroup.option("--history-name", type=str, help="Target history name")
@optgroup.group("History Options")
@optgroup.option("--ignore-case", "-i", is_flag=True, help="Ignore case when matching history names")
@optgroup.group("Galaxy Server Options")
@optgroup.option("--url", default="http://localhost:8080", help="URL of Galaxy instance")
@optgroup.option("--api-key", envvar="GALAXY_API_KEY", required=True, help="API key for Galaxy instance")
@optgroup.group("Upload Options")
@optgroup.option("--file-type", default="auto", type=str, help="Galaxy file type to use")
@optgroup.option("--dbkey", default="?", type=str, help="Genome Build for dataset")
@optgroup.option("--space-to-tab/--no-space-to-tab", default=False, help="Convert spaces to tabs")
@optgroup.option("--auto-decompress/--no-auto-decompress", default=True, help="Automatically decompress after upload")
@optgroup.option("--file-name", type=str, help="Filename to use in Galaxy history, if different from path")
@optgroup.group("General Options")
@optgroup.option("--storage", type=click.Path(), required=False, help="Store URLs to resume here")
@optgroup.option("--silent", is_flag=True, default=False, help="No output while uploading")
@optgroup.option("--debug", is_flag=True, default=False, help="Debug output")
@click.argument("path", type=click.Path(), nargs=-1)
def main(url, path, api_key, **kwargs):
    paths = path
    if not paths:
        message = "WARNING: No paths to upload specified (see --help)"
        click.echo(click.style(message, bold=True, fg="yellow"), err=True)
    elif len(paths) > 1 and kwargs["file_name"]:
        message = f"ERROR: --file-name option cannot be used with multiple paths"
        click.echo(click.style(message, bold=True, fg="red"), err=True)
        sys.exit(1)
    for path in paths:
        upload_file(url, path, api_key, **kwargs)


if __name__ == "__main__":
    main()


galaxy_upload/history.py
#!/usr/bin/env python3
"""Upload files to a Galaxy instance."""
import datetime
import re

import click
import rich.console
import rich.table
from bioblend.galaxy import GalaxyInstance


def make_table(quiet, histories=None):
    histories = histories or []
    # table setup
    if quiet:
        table = rich.table.Table(show_header=False, show_edge=False)
        table.add_column("ID", no_wrap=True)
    else:
        table = rich.table.Table(title="Active Histories")
        table.add_column("ID", style="cyan", no_wrap=True)
        table.add_column("Name", style="magenta")
        table.add_column("Last Modified", style="green")
    # add rows
    for history in histories:
        if quiet:
            row = (history["id"],)
        else:
            name = history["name"]
            update_time = datetime.datetime.fromisoformat(
                history["update_time"]
            ).strftime("%c")
            row = (
                history["id"],
                name,
                update_time,
            )
        table.add_row(*row)
    return table


def get_histories(gi, ignore_case=False, name=None):
    if name:
        if ignore_case:
            name_re = re.compile(name, flags=re.IGNORECASE)
        else:
            name_re = re.compile(name)

    # the "name" option to get_histories() only does exact matching, so we filter results ourselves
    histories = []
    for history in gi.histories.get_histories():
        if not name or name_re.search(history["name"]):
            histories.append(history)

    return histories


@click.command()
@click.option(
    "--url",
    envvar="GALAXY_URL",
    default="http://localhost:8080",
    help="URL of Galaxy instance",
)
@click.option(
    "--api-key",
    envvar="GALAXY_API_KEY",
    required=True,
    help="API key for Galaxy instance",
)
@click.option("--quiet", "-q", is_flag=True, help="Only output history IDs")
@click.option(
    "--ignore-case", "-i", is_flag=True, help="Ignore case when matching history names"
)
@click.argument("name", required=False)
def main(
    url,
    api_key,
    quiet,
    ignore_case,
    name,
):
    gi = GalaxyInstance(url, api_key)
    histories = get_histories(gi, ignore_case, name)
    table = make_table(quiet, histories=histories)
    console = rich.console.Console()
    console.print(table)


if __name__ == "__main__":
    main()


galaxy_upload/upload.py
#!/usr/bin/env python3
"""Upload files to a Galaxy instance."""
import os
import sys
import traceback

import click
import rich.console
import rich.progress
from bioblend import ConnectionError
from bioblend.galaxy import GalaxyInstance
from click_option_group import RequiredMutuallyExclusiveOptionGroup, optgroup
from tusclient.fingerprint import fingerprint

from .history import get_histories, make_table


def make_bar(file_name, total=None):
    columns = (
        rich.progress.TextColumn("[progress.description]{task.description}"),
        rich.progress.BarColumn(),
        rich.progress.DownloadColumn(),
        rich.progress.TransferSpeedColumn(),
        rich.progress.TextColumn("eta"),
        rich.progress.TimeRemainingColumn(),
    )

    bar = rich.progress.Progress(*columns)
    task_id = bar.add_task(file_name, total=total)
    return (bar, task_id)


def find_history(gi, history_name, ignore_case):
    histories = get_histories(gi, ignore_case=ignore_case, name=history_name)
    if not histories:
        message = f"ERROR: No histories matching {history_name} found!"
        click.echo(click.style(message, bold=True, fg="red"), err=True)
        sys.exit(1)
    elif len(histories) > 1:
        table = make_table(quiet=False, histories=histories)
        console = rich.console.Console()
        console.print(table)
        message = f"ERROR: Multiple histories matching {history_name} found! Use --history-id to select one."
        click.echo(click.style(message, bold=True, fg="red"), err=True)
        sys.exit(1)
    return histories[0]


def upload_file(
    url,
    path,
    api_key,
    history_id=None,
    history_name=None,
    ignore_case=False,
    file_type="auto",
    dbkey="?",
    space_to_tab=False,
    auto_decompress=False,
    file_name=None,
    storage=None,
    silent=False,
    debug=False,
):
    file_name = file_name or os.path.basename(path)

    gi = GalaxyInstance(url, api_key)

    upload_kwargs = {
        "file_type": file_type,
        "dbkey": dbkey,
        "file_name": file_name,
        "space_to_tab": space_to_tab,
        "auto_decompress": auto_decompress,
    }

    if not history_id:
        history = find_history(gi, history_name, ignore_case)
        history_id = history["id"]
        if debug:
            click.echo(
                f"History name search '{history_name}' resolved to history {history_id}: {history['name']}"
            )

    try:
        if silent:
            gi.tools.upload_file_tus(
                path,
                history_id,
                storage=storage,
                **upload_kwargs,
            )
            return

        uploader = gi.get_tus_uploader(
            path,
            storage=storage,
        )
        file_size = uploader.get_file_size()
        bar, task_id = make_bar(file_name, total=file_size)

        with bar:
            last_offset = 0
            while uploader.offset < file_size:
                uploader.upload_chunk()
                bar.update(task_id, advance=(uploader.offset - last_offset))
                last_offset = uploader.offset

        gi.tools.post_to_fetch(path, history_id, uploader.session_id, **upload_kwargs)

    except ConnectionError as exc:
        if exc.status_code == 404 and storage:
            fingerprinter = fingerprint.Fingerprint()
            with open(path, "rb") as fh:
                fp_hash = fingerprinter.get_fingerprint(fh)
            message = (
                f"Unable to resume, previous upload may have been removed from server (hint: remove {fp_hash}"
                f" from {storage} or change storage to reupload from the start: {exc}"
            )
        else:
            message = str(exc)
        message = f"ERROR: {message}"
        if debug:
            click.echo(traceback.format_exc(), nl=False)
        click.echo(click.style(message, bold=True, fg="red"), err=True)


@click.command()
@optgroup.group("History Selection Options", cls=RequiredMutuallyExclusiveOptionGroup)
@optgroup.option("--history-id", type=str, help="Target history ID")
@optgroup.option("--history-name", type=str, help="Target history name")
@optgroup.group("History Options")
@optgroup.option(
    "--ignore-case", "-i", is_flag=True, help="Ignore case when matching history names"
)
@optgroup.group("Galaxy Server Options")
@optgroup.option(
    "--url",
    envvar="GALAXY_URL",
    default="http://localhost:8080",
    help="URL of Galaxy instance",
)
@optgroup.option(
    "--api-key",
    envvar="GALAXY_API_KEY",
    required=True,
    help="API key for Galaxy instance",
)
@optgroup.group("Upload Options")
@optgroup.option(
    "--file-type", default="auto", type=str, help="Galaxy file type to use"
)
@optgroup.option("--dbkey", default="?", type=str, help="Genome Build for dataset")
@optgroup.option(
    "--space-to-tab/--no-space-to-tab", default=False, help="Convert spaces to tabs"
)
@optgroup.option(
    "--auto-decompress/--no-auto-decompress",
    default=True,
    help="Automatically decompress after upload",
)
@optgroup.option(
    "--file-name",
    type=str,
    help="Filename to use in Galaxy history, if different from path",
)
@optgroup.group("General Options")
@optgroup.option(
    "--storage", type=click.Path(), required=False, help="Store URLs to resume here"
)
@optgroup.option(
    "--silent", is_flag=True, default=False, help="No output while uploading"
)
@optgroup.option("--debug", is_flag=True, default=False, help="Debug output")
@click.argument("path", type=click.Path(exists=True, dir_okay=False), nargs=-1)
def main(url, path, api_key, **kwargs):
    paths = path
    if not paths:
        message = "WARNING: No paths to upload specified (see --help)"
        click.echo(click.style(message, bold=True, fg="yellow"), err=True)
    elif len(paths) > 1 and kwargs["file_name"]:
        message = "ERROR: --file-name option cannot be used with multiple paths"
        click.echo(click.style(message, bold=True, fg="red"), err=True)
        sys.exit(1)
    for path in paths:
        upload_file(url, path, api_key, **kwargs)


if __name__ == "__main__":
    main()
